File: packages/shared/src/schemas/live-session.ts
import { z } from 'zod';

export const LifecycleStateSchema = z.enum([
  'DEPLOYING',
  'RUNNING',
  'PAUSED',
  'STOPPING',
  'STOPPED',
]);

export const PositionSchema = z.object({
  instrument: z.string(),
  side: z.enum(['LONG', 'SHORT', 'FLAT']),
  quantity: z.string(),
  avg_entry_price: z.string(),
  unrealized_pnl: z.string(),
  realized_pnl: z.string(),
});

export const OrderSchema = z.object({
  id: z.string(),
  side: z.enum(['BUY', 'SELL']),
  type: z.enum(['MARKET', 'LIMIT', 'STOP_MARKET', 'STOP_LIMIT']),
  quantity: z.string(),
  price: z.string().optional(),
  fill_price: z.string().optional(),
  fill_quantity: z.string().optional(),
  status: z.enum(['PENDING', 'SUBMITTED', 'FILLED', 'PARTIALLY_FILLED', 'CANCELLED', 'REJECTED']).optional(),
});

export const PauseResumeEntrySchema = z.object({
  action: z.enum(['PAUSE', 'RESUME']),
  timestamp: z.string().datetime(),
  reason: z.string().optional(),
});

export const LiveDeploySchema = z.object({
  strategy_version_hash: z.string(),
  trading_pair: z.string().min(1),
  parameters: z.record(z.string(), z.unknown()).optional(),
  risk_profile_id: z.string().uuid(),
  confirmed_by_session: z.string(),
});

export const LiveSessionSchema = z.object({
  id: z.string().uuid(),
  strategy_version_hash: z.string(),
  trading_pair: z.string(),
  exchange: z.string().default('BINANCE'),
  lifecycle_state: LifecycleStateSchema,
  positions: z.array(PositionSchema),
  open_orders: z.array(OrderSchema),
  realized_pnl: z.string(),
  unrealized_pnl: z.string(),
  risk_profile_id: z.string().uuid(),
  parameters: z.record(z.string(), z.unknown()).optional(),
  confirmed_by: z.string(),
  started_at: z.string().datetime(),
  paused_at: z.string().datetime().nullable().optional(),
  stopped_at: z.string().datetime().nullable().optional(),
  pause_resume_history: z.array(PauseResumeEntrySchema).optional(),
});

export const LiveDeployResponseSchema = z.object({
  id: z.string().uuid(),
  lifecycle_state: LifecycleStateSchema,
  ws_url: z.string(),
});

export const KillSwitchResponseSchema = z.object({
  killed_sessions: z.number().int().nonnegative(),
  cancelled_orders: z.number().int().nonnegative(),
  flattened_positions: z.number().int().nonnegative(),
  executed_at: z.string().datetime(),
});

export type LifecycleState = z.infer<typeof LifecycleStateSchema>;
export type Position = z.infer<typeof PositionSchema>;
export type Order = z.infer<typeof OrderSchema>;
export type LiveDeploy = z.infer<typeof LiveDeploySchema>;
export type LiveSession = z.infer<typeof LiveSessionSchema>;
export type LiveDeployResponse = z.infer<typeof LiveDeployResponseSchema>;
export type KillSwitchResponse = z.infer<typeof KillSwitchResponseSchema>;

File: engine/src/schemas/live_session.py
"""Pydantic models for Live Session entity."""

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel


class LifecycleState(StrEnum):
    DEPLOYING = "DEPLOYING"
    RUNNING = "RUNNING"
    PAUSED = "PAUSED"
    STOPPING = "STOPPING"
    STOPPED = "STOPPED"


class Position(BaseModel):
    """Current position in a trading instrument."""
    instrument: str
    side: str  # LONG, SHORT, FLAT
    quantity: str
    avg_entry_price: str
    unrealized_pnl: str
    realized_pnl: str


class Order(BaseModel):
    """Order details."""
    id: str
    side: str  # BUY, SELL
    type: str  # MARKET, LIMIT, STOP_MARKET, STOP_LIMIT
    quantity: str
    price: str | None = None
    fill_price: str | None = None
    fill_quantity: str | None = None
    status: str | None = None


class PauseResumeEntry(BaseModel):
    """Record of a pause/resume action."""
    action: str  # PAUSE, RESUME
    timestamp: datetime
    reason: str | None = None


class LiveDeploy(BaseModel):
    """Request body for deploying a strategy to live trading."""
    strategy_version_hash: str
    trading_pair: str
    parameters: dict[str, Any] | None = None
    risk_profile_id: UUID
    confirmed_by_session: str


class LiveSession(BaseModel):
    """Full live session entity."""
    id: UUID
    strategy_version_hash: str
    trading_pair: str
    exchange: str = "BINANCE"
    lifecycle_state: LifecycleState
    positions: list[Position]
    open_orders: list[Order]
    realized_pnl: str
    unrealized_pnl: str
    risk_profile_id: UUID
    parameters: dict[str, Any] | None = None
    confirmed_by: str
    started_at: datetime
    paused_at: datetime | None = None
    stopped_at: datetime | None = None
    pause_resume_history: list[PauseResumeEntry] | None = None


class LiveDeployResponse(BaseModel):
    """Response after deploying a strategy."""
    id: UUID
    lifecycle_state: LifecycleState
    ws_url: str


class KillSwitchResponse(BaseModel):
    """Response after activating the kill switch."""
    killed_sessions: int
    cancelled_orders: int
    flattened_positions: int
    executed_at: datetime
